src/app/ClusterObjectUtils.h

```diff
@@ -0,0 +1,312 @@
+ *
+ */
+template <typename X, typename std::enable_if<std::is_class<X>::value, X>::type* = nullptr>
+CHIP_ERROR Decode(TLV::TLVReader &reader, X &x) {
+    return x.Decode(reader);
+}
+
+/*
+ * @brief
+ *
+ * This class provides an iteratable decoder of list items within TLV payloads
+ * such that no memory has to be provided ahead of time to store the entirety of the decoded
+ * list contents.
+ *
+ * Specifically, this provides a read-only input iterator that overloads the '->' operator to dynamically
+ * decode the list item at a given position in the stored TLVReader and return it to the caller.
+ *
+ */
+template <typename T>
+class IteratableList {
+public:
+    void SetReader(TLV::TLVReader& reader) { mReader = reader; }
+
+    size_t size() {
+        CHIP_ERROR err;
+        TLV::TLVReader reader;
+        reader.Init(mReader);
+        size_t count = 0;
+
+        while ((err = reader.Next()) == CHIP_NO_ERROR) {
+            count++; 
+        }
+
+        return count;
+    }
+
+    class Iterator {
+    public:
+        /*
+         * @brief
+         *
+         * Since neither the iterator constructor nor the operator overload permit sending errors back,
+         * encapsulate a tuple of the value and error that is then returned to the caller when decoding
+         */
+        struct ValuePair {
+            bool HasError() { return (status != CHIP_NO_ERROR); }
+            CHIP_ERROR status;
+            T value;
+        };
+
+        using iterator_category = std::input_iterator_tag;
+        using value_type = ValuePair;
+        using pointer = ValuePair*;
+        using reference = ValuePair&;
+
+        /*
+         * @brief
+         *
+         * Constructor
+         *
+         * reader       Pointer to a TLV reader whoose contents are copied into an internal
+         *              reader instance that is then used to track the iterator position 
+         *              within the backing buffer.
+         *
+         * signature    A unique signature that associated it with a given IteratableList instance
+         *              This permits actually comparing two iterators trivially to ensure they're
+         *              both logically iterating over the same list.
+         *
+         * isEnd        A flag that indicates if the iterator is pointing at the end of the TLV list.
+         *              This is preferable to actually iterating over the list till the end.
+         */
+        Iterator(TLV::TLVReader *reader, uintptr_t signature, bool isEnd = false) {
+            mSignature = signature;
+            mIsEnd = isEnd;
+            mValue.status = CHIP_NO_ERROR;
+            mDecoded = false;
+
+            if (!isEnd && reader) {
+                mReader.Init(*reader);
+                mValue.status = mReader.Next();
+            }
+        }
+
+        friend bool operator==(const Iterator& a, const Iterator& b) {
+            return ((a.mSignature == b.mSignature) && ((a.mIsEnd == b.mIsEnd)));
```

Comment:
How would we get into the situation where signatures match, but IsEnd doesn't?  Doesn't that mean that one or the other of the iterators has incorrect internal state?

Reply:
Have since pivoted away from this model.

Reply:
Thanks!  Will review.